src/Microsoft.Health.Fhir.Api/Controllers/AadSmartOnFhirProxyController.cs

```diff
@@ -124,21 +123,20 @@ public AadSmartOnFhirProxyController(IOptions<SecurityConfiguration> securityCon
                 launch = Base64UrlEncoder.Encode("{}");
             }
 
-            JObject newStateObj = JObject.Parse("{}");
-            newStateObj.Add("s", state);
-            newStateObj.Add("l", launch);
+            var newStateObj = new JObject
+            {
+                { "s", state },
+                { "l", launch },
+            };
 
-            string newState = Base64UrlEncoder.Encode(newStateObj.ToString(Newtonsoft.Json.Formatting.None));
+            string newState = Base64UrlEncoder.Encode(newStateObj.ToString());
 
-            Uri callbackUrl = new Uri(
-                Request.Scheme + "://" + Request.Host + "/AadSmartOnFhirProxy/callback/" +
-                Base64UrlEncoder.Encode(redirectUri.ToString()));
+            var callbackUrl = UriHelper.BuildAbsolute(Request.Scheme, Request.Host, "/AadSmartOnFhirProxy/callback", $"/{Base64UrlEncoder.Encode(redirectUri.ToString())}");
 
-            StringBuilder queryStringBuilder = new StringBuilder();
-            queryStringBuilder.Append($"response_type={responseType}&redirect_uri={callbackUrl.ToString()}&client_id={clientId}");
+            var queryStringBuilder = new StringBuilder($"response_type={HttpUtility.UrlEncode(responseType)}&redirect_uri={callbackUrl}&client_id={HttpUtility.UrlEncode(clientId)}");
```

Comment:
Have you considered using QueryBuilder?

```
            var queryBuilder = new QueryBuilder()
            {
                { "response_type", responseType },
                { "redirect_uri", callbackUrl },
                { "client_id", clientId },
                { "state", newState },
            };

            if (!_isAadV2)
            {
                queryBuilder.Add("resource", aud);
            }
```
then use `queryBuilder.ToString()` automatically encodes the values I believe. Though we do seem to use both UrlEncode and EscapeDataString so I am not sure if it will affect that.